### PR TITLE
concept: benchmarking: initial benchmarking index

### PR DESCRIPTION
**Pull request by amx-piap**

### CODE REVIEW

**Comment by BorjaOuterelo** [comment]
@amx-piap there are some links with issues:
You can check the CI result in https://gitlab.com/micro-ROS/ci/micro-ROS.github.io/-/jobs

> 
> - ./_site/docs/concepts/benchmarking/index.html
>   *  linking to internal hash #shadow-builder that does not exist (line 588)
>      <a href="#shadow-builder">Shadow Builder</a>
>   *  linking to internal hash #step-to-start-bencharking that does not exist (line 590)
>      <a href="#step-to-start-bencharking">Step to start Benchmarking</a>
>   *  linking to internal hash #tfa-trace-framework-abstraction that does not exist (line 587)
>      <a href="#tfa-trace-framework-abstraction">TFA - Trace Framework Abstraction</a>

File: _docs/concepts/benchmarking/index.md [hunk 1]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+- [Our benchmarking tool framework](#our-benchmarking-tool-framework)
+- [Binary generation for instrumented code](#binary-generation-for-instrumented-code)
+- [TFA - Trace Framework Abstraction](#tfa-trace-framework-abstraction)
+- [Shadow Builder](#shadow-builder)
+- [Binary generation for instrumented code](#binary-generation-for-instrumented-code)
+- [Step to start Benchmarking](#step-to-start-bencharking)
+
+Developing working and stable application from the scribbles to the final
+executing binary is long and hard tasks. During this process developers may come
+across stabilities issues, perfomances issues. In addition to these issues, some
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
across stability issues, performance issues. In addition to these issues, some
```

File: _docs/concepts/benchmarking/index.md [hunk 2]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+specified QoS might be difficul to quantify.  Solving those problems without the
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
specified QoS might be difficult to quantify.  Solving those problems without the
```

File: _docs/concepts/benchmarking/index.md [hunk 3]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+proper tools might be a frustrating , tedious tasks leading to reduce developers
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
proper tools might be frustrating , tedious tasks leading to reduce developers
```

File: _docs/concepts/benchmarking/index.md [hunk 4]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+efficiency. An adapted benchmarking tool could overcome all those development
+obstacles and increase development time.  There are different KPI (Keep
+Performance Indicators) that one might be interested into. In the framework of
+this micro-ROS, the KPI can be freely chosen by the developer. In this way, the
+benchmarking tool will remain flexible and allow the community to constantly add
+some support for a lot of different KPI.
+
+The problems we want to tackle are: 
+
+ * Out there, many benchmarking tools exists. Each of targeting different KPIs.
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
 * Out there, many benchmarking tools exist. Each of targeting different KPIs. 
```

File: _docs/concepts/benchmarking/index.md [hunk 5]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+ * Too few time/resources to code benchmarking tool for each.
+ * Avoid code overhead: Keep code clarity.
+ * Avoid execution overhead: Do not want to make execution slower when benchmarking.
+
+## Our Benchmarking tool framework
+
+The benchmarking tool under development is providing a framework to allow
+developers to create their own benchmarking tool. Each part a developer wants to
+benchmark can be added as a plugin using the provided framework. In this way
+plugins can be shared and this improve re-usability as much as possible.
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
plugins can be shared and this improves re-usability as much as possible.
```

File: _docs/concepts/benchmarking/index.md [hunk 6]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+
+
+## Binary generation for instrumented code 
+
+The binary generation is the process of compiling the source code. In order to
+benchmark, previously to compile the source code, it is necessary to instrument
+the code. The code will instrumented in a transparent way for the
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
the code. The code will be instrumented in a transparent way for the
```

File: _docs/concepts/benchmarking/index.md [hunk 7]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+programmer/user. Therefore, a configuration file provided by the programmer will
+be parsed and code inject as described in a configuration file.
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
be parsed and code injected as described in a configuration file. 
```

File: _docs/concepts/benchmarking/index.md [hunk 8]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+
+### Step 1 - Binary Generation : Receiving inputs
+
+The binary generation's pipeline receives two inputs to work with:
+ * Configuration Benchmarking file.
+ * Source code to benchmark.
+
+In short, the configuration describes:
+
+ * What is bench-marked (sources).
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
 * What is benchmarked (sources).
```

File: _docs/concepts/benchmarking/index.md [hunk 9]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+ * The different benchmarking to be achieved.
+ * The targeted platforms.
+
+In addition to parsing, the Shadow Builder is in charge of checking
+capabilities and consistency within the configuration file and the different
+TFA's plugins registered in the TFA module.
+
+### Step 3 - Binary Generation: TFA Execution
+
+Once parsed and check against the TFA module capabilities, the Shadow
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
Once parsed and checked against the TFA module capabilities, the Shadow
```

File: _docs/concepts/benchmarking/index.md [hunk 10]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+Builder will be in charge of translating configuration into source code. The
+translated sources will be achieved also in cooperation with the TFA module. The
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
translated sources will also be achieved in cooperation with the TFA module. The
```

File: _docs/concepts/benchmarking/index.md [hunk 11]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+detailed steps of the TFA can be found here. At the end of this step, the TFA
+will generate the new forged source code ready for compilation. In addition to
+patched source code, the TFA will generate scripts that will the benchmarks.
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]: > the TFA will generate scripts that wil...

include?
provide?

File: _docs/concepts/benchmarking/index.md [hunk 12]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+
+### Step 4 - Binary Generation: Compilation
+
+The compilation will happen for every kind of benchmarkings and
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
The compilation will happen for every kind of benchmarks and
```

File: _docs/concepts/benchmarking/index.md [hunk 13]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+builder. The shadow builder will retrieve capabilities of the plugins and
+request from the developer, match them and generated software according to the
+matches.
+
+## TFA - Trace Framework Abstraction
+
+The Shadow builder alone only parse comments from the application and pass it
+along to the Trace Framework Abstraction (TFA) Core. The TFA core is aware of
+the plugins that are available, all the plugins’ capabilities and platform
+target. The process goes a explained below:
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
target. The process goes as explained below: 
```

File: _docs/concepts/benchmarking/index.md [hunk 14]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+## TFA - Trace Framework Abstraction
+
+The Shadow builder alone only parse comments from the application and pass it
+along to the Trace Framework Abstraction (TFA) Core. The TFA core is aware of
+the plugins that are available, all the plugins’ capabilities and platform
+target. The process goes a explained below: 
+
+ * The line containing the functionality Benchmarking::XX::YY will be checked
+   against all the available plugins.
+ * Plugin that are capable to handle functionality will respond with a piece of
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
 * Plugin that is capable of handling functionality will respond with a piece of
```

File: _docs/concepts/benchmarking/index.md [hunk 15]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+   code that will be replaced with a piece of code.
+ * Then the output file will be added in a folder corresponding to the platform
+   type and benchmarking type.
+
+Being generic is the key for this benchmarking tool. The plugins will in
+contrary bring the specific implementation needed to benchmark  a specific
+platform. Every plugins will to provide information as  requested by the parser:
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
platform. Every plugin will provide information as  requested by the parser:
```

File: _docs/concepts/benchmarking/index.md [hunk 16]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+
+ * Provide list of supported platform.
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
 * Provide a list of the supported platform. 
```

File: _docs/concepts/benchmarking/index.md [hunk 17]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+   type and benchmarking type.
+
+Being generic is the key for this benchmarking tool. The plugins will in
+contrary bring the specific implementation needed to benchmark  a specific
+platform. Every plugins will to provide information as  requested by the parser:
+
+ * Provide list of supported platform. 
+ * Provide a list of functions that are handled.
+ * Provide snippets codes that will be added for benchmarking.
+ * Provide list of patches and/or patch code
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
 * Provide a list of patches and/or patch code
```

File: _docs/concepts/benchmarking/index.md [hunk 18]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+ * Optional provide an end script to run andi execute the benchmarks
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
 * Optional provide an end script to run and execute the benchmarks
```

File: _docs/concepts/benchmarking/index.md [hunk 19]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+## SB - Shadow Builder
+
+This section will introduce some concept related to the shadow builder (SB).
+
+The Shadow builder is a tool that will transparently instrument the code to
+benchmark. The tools will be able to output an “instrumented code” that will be
+later be compiled as a normal code. The following steps describe what the shadow
+builder process flow:
+
+ * Get configuration file from user (Benchmarking Configuration File).
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
 * Get configuration file from the user (Benchmarking Configuration File).
```

File: _docs/concepts/benchmarking/index.md [hunk 20]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+
+The SB (Shadow Builder) is meant to be as transparent as possible for the user.
+And if the benchmarking is not activated, it should be bypassed.  Get
+appropriate sources
+
+The SB is in charge of getting the path/git repository to the source code that
+needs to be benchmarking. The benchmarking. The sources are specified by the
+user in the benchmarking configuration file.  Injecting code
+
+In order to inject code, there are some tool that allow this. CLang AST tool
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
In order to inject code, there are some tools that allow this. CLang AST tool
```

File: _docs/concepts/benchmarking/index.md [hunk 21]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+will allow to inject some code. 
+
+## Step to start Benchmarking
+
+The shadow Builder will be executed as follow:
+
+ * Software sources are passed to the Shadow Builder.
+ * The source are passed and upon comments containing /*Benchmarking::XX::YY*/
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]:
```suggestion
 * The source is parsed and upon comments containing /*Benchmarking::XX::YY*/
```

File: _docs/concepts/benchmarking/index.md [hunk 22]
```diff
@@ -3,4 +3,159 @@ title: Introduction to Benchmarking
...
+ * Get appropriate sources.
+ * Execute Trace Framework Abstraction Configuration file.
+ * Parse the sources file needed Injecting code.
+ * Compile the targeted binary for different platform.
+ * If needed, depending what type benchmark is undertaken, compile another
+   target binary benchmarking.
+
+The SB (Shadow Builder) is meant to be as transparent as possible for the user.
+And if the benchmarking is not activated, it should be bypassed.  Get
+appropriate sources
```
Comment (BorjaOuterelo) [inline _docs/concepts/benchmarking/index.md]: It looks like this is some text not meant to be here.

**Comment by BorjaOuterelo** [comment]
@amx-piap, @tkolcon any update on this PR is still valid?

**Comment by polonovski** [comment]
I mean that was the part of the website we had to integrate. There was some issue? I am not sure anymore that was 2.5 mounth ago...

But it had to be integrated.

**Comment by polonovski** [comment]
So yes it is valid

**Comment by BorjaOuterelo** [comment]
@polonovski I will rebase it in the current master and push here if thats ok for you.

**Comment by amx-piap** [comment]
Sure thank you

**Comment by BorjaOuterelo** [comment]
@amx-piap It is rebased and added some language changes. Can we merge it?.

File: _docs/concepts/benchmarking/index.md (hunk 11 above)
Comment (amx-piap) [reply _docs/concepts/benchmarking/index.md]: that can be used to launch benchmarks.

**Comment by amx-piap** [comment]
Yes we can merge it thanks.

**Pull request merged**